src/compiler/checker.ts

```diff
@@ -4436,17 +4517,20 @@ namespace ts {
                 // it may hold in a structural comparison.
                 // Report structural errors only if we haven't reported any errors yet
                 let reportStructuralErrors = reportErrors && errorInfo === saveErrorInfo;
-                // identity relation does not use apparent type
+                // Identity relation does not use apparent type
                 let sourceOrApparentType = relation === identityRelation ? source : getApparentType(source);
-                if (sourceOrApparentType.flags & TypeFlags.ObjectType && target.flags & TypeFlags.ObjectType) {
+                // In a check of the form X = A & B, we will have previously checked if A relates to X or B relates
+                // to X. Failing both of those we want to check if the aggregation of A and B's members structurally
+                // relates to X. Thus, we include intersection types on the source side here.
+                if (sourceOrApparentType.flags & (TypeFlags.ObjectType | TypeFlags.Intersection) && target.flags & TypeFlags.ObjectType) {

```

Comment:
I have been thinking a bit more about this `TypeFlags.Intersection` check, and I think it is too permissive because the intersection could have unions as constituents. Here is an example:
```ts
interface Target {
    x?: number;
}
interface A1 {
    a: any;
    x: string;
}
interface A2 {
    a: any;
    x2: string;
}
interface B1 {
    b: any;
    x: string;
}
interface B2 {
    b: any;
    x2: string;
}

var t: Target;
var s: (A1 | A2) & (B1 | B2);
t = s;
```
This assignment is supposed to fail, because no matter which side of the intersection you're looking it, there could be a union constituent where x has the wrong type (string instead of number). But because you are recursively diving into the properties of the unions as well as the intersections, you end up losing information about the property x. As a result, we allow the assignment, even though it is bad.